### PR TITLE
Add unittest for testing request-handling

### PR DESCRIPTION
**Pull request by jeffreymeetkai**
- Test edge cases error raising
- Test response creation for non-streaming
- Fix discovered small bugs
  - `tool_calls` returns `None` instead of empty string now
  - inference handles `request.function_call` if it is not None

### CODE REVIEW

File: tests/test_request_handling.py [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+        self.default_tools = [
+            {"type": "function", "function": self.default_functions[0]}
+        ]
+        self.v2_tokenizer = AutoTokenizer.from_pretrained(
+            "meetkai/functionary-small-v2.4"
+        )
+        self.v2_llama3_tokenizer = AutoTokenizer.from_pretrained(
+            "meetkai/functionary-small-v2.5"
+        )
+        self.default_text_str = "Normal text generation"
```
[the remark this answers is not in the document]
Comment (jeffreymeetkai) [reply tests/test_request_handling.py:146]: I think there's no need for now because it is only default_text_str, default_tool_call_name and default_tool_call_args only. Maybe when we want to test other more extensive cases in the future which require more data, we can consider this?

File: tests/test_request_handling.py [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+                self.assertEqual(
+                    ChatMessage(**chat_mess),
+                    test_case["expected_result"],
+                    f"Wrong ChatMessage for version: {prompt_template.version} | test: `{test_case['test_aim']}`",
+                )
+                self.assertEqual(
+                    finish_reason,
+                    test_case["expected_finish_reason"],
+                    f"Wrong finish reason for version: {prompt_template.version} | test: `{test_case['test_aim']}`",
+                )
```
[the remark this answers is not in the document]
Comment (jeffreymeetkai) [reply tests/test_request_handling.py:532]: Yes, I will include testing streaming in a future PR.

File: tests/test_request_handling.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+            if prompt_template.version == "v2":
+                if tool_func_choice == "none":
+                    raw_response = raw_response[len("all\n<|content|>") :]
+                elif isinstance(tool_func_choice, Tool) or isinstance(
+                    tool_func_choice, Function
+                ):
+                    raw_response = raw_response[
+                        len(f"{self.default_tool_call_name}\n<|content|>") :
+                    ]
+            elif prompt_template.version == "v2.llama3":
```
Comment (khai-meetkai) [inline tests/test_request_handling.py:482]: I mean we don't need to check if prompt_template_version == a specific_version. If not when we add a new template (for example template of Phi3), we have to write if-else here. I think we can implement this without if-else for prompt_template.version using abstract methods:
+ convert_message_to_prompt
+ get_force_text_generation_prefix
+ get_force_function_call_prefix

File: tests/test_request_handling.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply tests/test_request_handling.py:148]: Yeah, it's up to you

File: functionary/prompt_template/base_template.py [hunk 4]
```diff
@@ -227,6 +228,38 @@ def get_force_function_call_prefix(self, function_name: str):
         """
         raise NotImplementedError
 
+    def get_raw_response_from_assistant_message(
+        self,
+        message: Dict[str, str],
```
Comment (khai-meetkai) [inline functionary/prompt_template/base_template.py:233]: Dict[str, Any]

File: tests/test_request_handling.py [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+                        }
+                    },
+                    "required": ["location"],
+                },
+            },
+        ]
+        self.default_tools = [
+            {"type": "function", "function": self.default_functions[0]}
+        ]
+        self.tokenizers_to_test = [
```
Comment (khai-meetkai) [inline tests/test_request_handling.py:140]: Actually I think tokenizers here are used to create prompt_template, right ? so why don't we just use 

self.test_template_versions = get_available_templates_versions() --> list of string in prompte_template/__init__.py 
so we don't need to care about downloading tokenizer from HF and also private model that we don't set public?

**Review by khai-meetkai — APPROVED** [review]
